Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,354 @@
+
+  /// Maximum number of pings that can be sent when there is no data/header frame to be sent.
+  private let maximumPingsWithoutData: UInt
+
+  /// If there are no data/header frames being received:
+  /// The minimum amount of time to wait between successive pings.
+  private let minimumSentPingIntervalWithoutData: TimeAmount
+
+  /// If there are no data/header frames being sent:
+  /// The minimum amount of time expected between receiving successive pings.
+  /// If the time between successive pings is less than this value, then the ping will be considered a bad ping from the peer.
+  /// Such a ping counts as a "ping strike".
+  /// Ping strikes are only applicable to server handler
+  private let minimumReceivedPingIntervalWithoutData: TimeAmount?
+
+  /// Maximum number of bad pings that the server will tolerate before sending an HTTP2 GOAWAY frame and closing the connection.
+  /// Setting it to `0` allows the server to accept any number of bad pings.
+  /// Ping strikes are only applicable to server handler
+  private let maximumPingStrikes: UInt?
+
+  /// When the last ping was received
+  private var lastReceivedPingDate: NIODeadline? = nil
+
+  /// When the last ping was sent
+  private var lastSentPingDate: NIODeadline? = nil
+
+  /// The number of pings sent on the transport without any data
+  private var sentPingsWithoutData = 0
+
+  /// Number of strikes
+  private var pingStrikes: UInt = 0
+
+  /// The scheduled task which will close the connection.
+  private var scheduledClose: Scheduled<Void>? = nil
+
+  /// Number of active streams
+  private var activeStreams = 0 {
+    didSet {
+      if activeStreams > 0 {
+        self.sentPingsWithoutData = 0
+      }
+    }
+  }
+
+  private let goAwayFrame = HTTP2Frame.FramePayload.goAway(lastStreamID: .rootStream, errorCode: .enhanceYourCalm, opaqueData: nil)
+
+  // For testing only
+  var testingOnlyNow: NIODeadline?
+
+  enum Action {
+    case none
+    case schedulePing(delay: TimeAmount, timeout: TimeAmount)
+    case cancelScheduledPing
+    case cancelScheduledTimeout
+    case reply(HTTP2Frame.FramePayload)
+  }
+
+  init(
+    pingCode: UInt64,
+    interval: TimeAmount,
+    timeout: TimeAmount,
+    permitWithoutCalls: Bool,
+    maximumPingsWithoutData: UInt,
+    minimumSentPingIntervalWithoutData: TimeAmount,
+    minimumReceivedPingIntervalWithoutData: TimeAmount? = nil,
+    maximumPingStrikes: UInt? = nil
+  ) {
+    self.pingCode = pingCode
+    self.interval = interval
+    self.timeout = timeout
+    self.permitWithoutCalls = permitWithoutCalls
+    self.maximumPingsWithoutData = maximumPingsWithoutData
+    self.minimumSentPingIntervalWithoutData = minimumSentPingIntervalWithoutData
+    self.minimumReceivedPingIntervalWithoutData = minimumReceivedPingIntervalWithoutData
+    self.maximumPingStrikes = maximumPingStrikes
+  }
+
+  func streamCreated() -> Action {
+    self.activeStreams += 1
+
+    if self.activeStreams == 1 {
```

Comment:
The choice of `1` here is interesting: can you elaborate on why we schedule the ping that way?

Reply:
Changed and now based on `startedAt`. If it's `nil`, I start the ping schedule task.